### PR TITLE
[mxnet] [build] Mxnet 1 9 mkl blas

### PR DESCRIPTION
**Pull request by ydaiming**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description

Update MXNet 1.9 inference CPU with MKL BLAS support.

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: mxnet/inference/docker/1.9/py3/Dockerfile.cpu [hunk 1]
```diff
@@ -79,6 +83,20 @@ RUN apt-get update \
 
 WORKDIR /
 
+RUN curl -L -o ~/mambaforge.sh https://github.com/conda-forge/miniforge/releases/download/${MAMBA_VERSION}/Mambaforge-${MAMBA_VERSION}-Linux-x86_64.sh \
```
Comment (shantanutrip) [inline mxnet/inference/docker/1.9/py3/Dockerfile.cpu:86]: When we pull the container and type `conda` it fails to work. This is leading to a few test failures. 

Can we fix this? We can refer to other dockerfiles that have conda setup, for eg. https://github.com/aws/deep-learning-containers/blob/master/pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu#L159-L186

File: mxnet/inference/docker/1.9/py3/Dockerfile.cpu [hunk 2]
```diff
@@ -79,6 +83,20 @@ RUN apt-get update \
...
+    mkl \
+    mkl-include \
+ # Upstream conda looks to have moved to 4.13 which is incompatible with mamba 0.22.1 and will fail the conda-forge installs.
+ # having "conda update conda" before the "conda -c conda-forge" commands will automatically update conda to 4.13.
+ # Moving conda update conda" after the "conda -c conda-forge" commands keep conda at 4.12 but will update other packages using
+ # the current conda 4.12
+ && /opt/conda/bin/conda update -y conda \
+ && /opt/conda/bin/conda clean -ya
+
 RUN pip install --upgrade pip --trusted-host pypi.org --trusted-host files.pythonhosted.org \
```
Comment (shantanutrip) [inline mxnet/inference/docker/1.9/py3/Dockerfile.cpu:100]: Do we need a `--no-cache-dir`

File: mxnet/inference/docker/1.9/py3/Dockerfile.cpu (hunk 1 above)
Comment (matherit) [reply mxnet/inference/docker/1.9/py3/Dockerfile.cpu:87]: Added /opt/conda/bin to path. Should be resolved.

File: mxnet/inference/docker/1.9/py3/Dockerfile.cpu (hunk 2 above)
Comment (matherit) [reply mxnet/inference/docker/1.9/py3/Dockerfile.cpu:101]: This does not appear to be a thing for conda install. https://conda.io/projects/conda/en/latest/commands/install.html

conda clean -ya should cleanup whatever is cached anyways.

File: mxnet/inference/docker/1.9/py3/Dockerfile.cpu [hunk 3]
```diff
@@ -173,7 +194,8 @@ RUN HOME_DIR=/root \
  && chmod +x /usr/local/bin/testOSSCompliance \
  && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
  && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} python \
- && rm -rf ${HOME_DIR}/oss_compliance*
+ && rm -rf ${HOME_DIR}/oss_compliance* \
+ && rm -iRf /root/.cache
```
Comment (shantanutrip) [inline mxnet/inference/docker/1.9/py3/Dockerfile.cpu:198]: Do we need to add this for the ec2 image as well? I see test failures on ec2 which might get resolved by adding this.

Comment (matherit) [reply mxnet/inference/docker/1.9/py3/Dockerfile.cpu:199]: I pushed a fix for this.

File: mxnet/training/docker/1.9/py3/cu112/Dockerfile.gpu [hunk 4]
```diff
@@ -42,7 +42,8 @@ ENV PYTHONDONTWRITEBYTECODE=1 \
     DGLBACKEND=mxnet \
     CUDNN_VERSION=8.1.0.77-1+cuda11.2 \
     NCCL_VERSION=2.8.3-1+cuda11.2 \
-    DEBIAN_FRONTEND=noninteractive
+    DEBIAN_FRONTEND=noninteractive \
+    DLC_CONTAINER_TYPE=training
```
Comment (shantanutrip) [inline mxnet/training/docker/1.9/py3/cu112/Dockerfile.gpu:46]: Why do we have training images in this PR?

File: mxnet/inference/docker/1.9/py3/cu112/Dockerfile.gpu [hunk 5]
```diff
@@ -31,7 +31,7 @@ LABEL dlc_major_version="1"
 
 ARG PYTHON_VERSION=3.8.10
 ARG MMS_VERSION=1.1.8
-ARG MX_URL=https://aws-mx-pypi.s3.us-west-2.amazonaws.com/1.9.0/aws_mx_cu112-1.9.0b20211227.163-py3-none-manylinux2014_x86_64.whl
+ARG MX_URL=https://framework-binaries.s3.us-west-2.amazonaws.com/mxnet/v1.9.0_telemetry/aws_mx_cu112-1.9.0b20221123.1-py3-none-manylinux2014_x86_64.whl
```
Comment (shantanutrip) [inline mxnet/inference/docker/1.9/py3/cu112/Dockerfile.gpu:34]: I see that the new binaries have been built in a different s3 bucket. Is there a reason for the same?

File: mxnet/inference/docker/1.9/py3/Dockerfile.cpu [hunk 6]
```diff
@@ -40,7 +42,11 @@ ENV PYTHONDONTWRITEBYTECODE=1 \
     LANG=C.UTF-8 \
     LC_ALL=C.UTF-8 \
     TEMP=/home/model-server/tmp \
-    DEBIAN_FRONTEND=noninteractive
+    DEBIAN_FRONTEND=noninteractive \
+    DLC_CONTAINER_TYPE=inference
```
Comment (shantanutrip) [inline mxnet/inference/docker/1.9/py3/Dockerfile.cpu:46]: What is the reason for adding this `env` variable?

File: test/dlc_tests/sanity/test_mxnet_blasmkl.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    run_cmd_on_container,
+    start_container,
+    stop_and_remove_container,
+)
+
+@pytest.mark.usefixtures("sagemaker")
+@pytest.mark.model("N/A")
+@pytest.mark.processor("cpu")
+@pytest.mark.integration("mxnet_blasmkl_sanity")
+def test_mxnet_blasmkl_sanity(mxnet_inference):
```
Comment (shantanutrip) [inline test/dlc_tests/sanity/test_mxnet_blasmkl.py:19]: Thanks for adding the test.

nit: do we want to call it `test_mxnet_blas_mkl_sanity` instead of `test_mxnet_blasmkl_sanity`?

File: mxnet/training/docker/1.9/py3/cu112/Dockerfile.gpu (hunk 4 above)
Comment (matherit) [reply mxnet/training/docker/1.9/py3/cu112/Dockerfile.gpu:46]: I brought in the commits from this PR to resolve some of the test failures.
https://github.com/aws/deep-learning-containers/pull/2458

If we merge this PR, we can just close the other and knock out both fixes in one go.

If this isn't something you'd like to do, feel free to revert and we can look at merging the other PR as well.

File: mxnet/inference/docker/1.9/py3/Dockerfile.cpu (hunk 6 above)
Comment (matherit) [reply mxnet/inference/docker/1.9/py3/Dockerfile.cpu:46]: This resolves the issue that mxnet inference images also report as training images. This enables the MXNet binary to know what type of image it is operating in.

File: test/dlc_tests/sanity/test_mxnet_blasmkl.py (hunk 7 above)
Comment (matherit) [reply test/dlc_tests/sanity/test_mxnet_blasmkl.py]: Resolved.

File: mxnet/training/docker/1.9/py3/cu112/Dockerfile.gpu (hunk 4 above)
Comment (shantanutrip) [reply mxnet/training/docker/1.9/py3/cu112/Dockerfile.gpu:46]: Can we revert all the training Dockerfile changes and handle them in a different PR? I say this because we have not run any tests on the training till now.

File: mxnet/inference/docker/1.9/py3/cu112/Dockerfile.gpu (hunk 5 above)
Comment (matherit) [reply mxnet/inference/docker/1.9/py3/cu112/Dockerfile.gpu:34]: We've been using this new bucket for awhile now. 
https://github.com/aws/deep-learning-containers/commit/db6977c607cc85f0af65ecd78a537a41e513dc34#diff-fe0edbe322492286efa96fdf55c6133c0c9cfd103f0b7406a775945cf5448bdf

It's also used for TF. https://github.com/aws/deep-learning-containers/blob/master/tensorflow/training/docker/2.11/py3/Dockerfile.cpu#L192

File: test/dlc_tests/sanity/test_mxnet_blas_mkl.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    ctx = Context()
+    container_name = get_container_name("mxnet-blas-mkl", mxnet_inference)
+    start_container(container_name, mxnet_inference, ctx)
+
+    output = run_cmd_on_container(
+        container_name, ctx, 'import mxnet; assert mxnet.runtime.Features().is_enabled("BLAS_MKL") == True', executable="python"
+    )
+
+    # If BLAS MKL is enabled, output should be blank. Otherwise, assertion errors will be raised.
+    assert f"{output.stdout}" == ""
```
Comment (shantanutrip) [inline test/dlc_tests/sanity/test_mxnet_blas_mkl.py:43]: If this assert condition fails, will we be able to `stop_and_remove_container` ? If not, can we have `try` and `finally` here?

File: test/dlc_tests/sanity/test_mxnet_blas_mkl.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    :param mxnet_inference: framework fixture for mxnet inference
+    """
+    _, framework_version = get_framework_and_version_from_tag(mxnet_inference)
+    if Version(framework_version) < Version("1.9.0"):
+        pytest.skip("Skipping this test MXNet versions less than 1.9.0 which do not use BLAS MKL.")
+    if "eia" in mxnet_inference:
+        pytest.skip("This test does not apply to EIA images.")
+    if "neuron" in mxnet_inference:
+        pytest.skip("Skipping because this is not relevant to Neuron images.")
+    if get_processor_from_image_uri(mxnet_inference) != "cpu":
```
[the remark this answers is not in the document]
Comment (matherit) [reply test/dlc_tests/sanity/test_mxnet_blas_mkl.py]: Done

File: test/dlc_tests/sanity/test_mxnet_blas_mkl.py (hunk 8 above)
Comment (matherit) [reply test/dlc_tests/sanity/test_mxnet_blas_mkl.py]: Done

File: test/dlc_tests/sanity/test_mxnet_blas_mkl.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    Check that the container's version of MXNet includes BLAS MKL.
+    :param mxnet_inference: framework fixture for mxnet inference
+    """
+    _, framework_version = get_framework_and_version_from_tag(mxnet_inference)
+    if Version(framework_version) < Version("1.9.0"):
+        pytest.skip("Skipping this test MXNet versions less than 1.9.0 which do not use BLAS MKL.")
+    if "eia" in mxnet_inference:
+        pytest.skip("This test does not apply to EIA images.")
+    if "neuron" in mxnet_inference:
+        pytest.skip("Skipping because this is not relevant to Neuron images.")
```
[the remark this answers is not in the document]
Comment (matherit) [reply test/dlc_tests/sanity/test_mxnet_blas_mkl.py]: I was following this test's example: https://github.com/aws/deep-learning-containers/blob/master/test/dlc_tests/sanity/test_torchvision.py#L51

I went ahead and removed them for now.

**Review by shantanutrip — DISMISSED** [review]
Please revert the toml file.

**Review by kace — DISMISSED** [review]

**Review by shantanutrip — APPROVED** [review]

**Review by kace — APPROVED** [review]
